packages/graphql/src/components/DeferredInterfaceValidation.ts

```diff
@@ -0,0 +1,245 @@
+  }
+
+  return fields;
+}
+
+/**
+ * Register a type for deferred interface implementation validation.
+ * Resolves interface refkeys immediately while we're still in the render context with access to the binder.
+ * Field extraction is deferred until validation time to ensure all members have been added.
+ */
+export function registerForValidation(
+  typeName: string,
+  symbol: GraphQLOutputSymbol,
+  interfaces: Children[],
+): void {
+  // Resolve interfaces immediately while we have binder context, including transitive ones
+  const resolvedInterfaces: Record<string, ResolvedInterface> = {};
+
+  function resolveInterfacesRecursively(interfaces: Children[]) {
+    for (const iface of interfaces) {
+      // An interface can be either a refkey or a string literal, only validate refkey references,
+      // as string literal interfaces are already validated and rendered as is.
+      if (isRefkey(iface)) {
+        const reference = ref(iface);
+        const [, ifaceSymbol] = reference();
+        const nameStr = String(ifaceSymbol?.name);
+
+        if (ifaceSymbol && !(nameStr in resolvedInterfaces)) {
+          resolvedInterfaces[nameStr] = { symbol: ifaceSymbol };
+
+          // Recursively resolve parent interfaces
+          const parentInterfaces = ifaceSymbol.metadata.implements as
+            | Children[]
+            | undefined;
+          if (parentInterfaces?.length) {
+            resolveInterfacesRecursively(parentInterfaces);
+          }
+        }
+      }
+    }
+  }
+
+  resolveInterfacesRecursively(interfaces);
+
+  pendingValidations.push({
+    typeName,
+    typeSymbol: symbol,
+    interfaces: Object.values(resolvedInterfaces),
+  });
+}
+
+/**
+ * Run all pending interface implementation validations
+ *
+ * This is called automatically by the SourceFile component after rendering is complete.
+ * Users don't need to call this manually - it's handled by the framework.
+ *
+ * Errors are collected rather than thrown, and can be retrieved with getValidationErrors().
+ */
+export function runPendingValidations(): void {
+  const validations = pendingValidations;
+  pendingValidations = [];
+  validationErrors = [];
+
+  for (const { typeName, typeSymbol, interfaces } of validations) {
+    // Extract field information now that all members have been added
+    const typeFields = extractFieldInfo(typeSymbol);
+
+    try {
+      // Validate each resolved interface
+      for (const { symbol: interfaceSymbol } of interfaces) {
+        validateTypeImplementsInterface(typeName, typeFields, interfaceSymbol);
+      }
+    } catch (e) {
+      if (e instanceof Error) {
+        validationErrors.push(e);
+      }
+    }
+  }
+}
+
+/**
+ * Validates that a type correctly implements a single interface
+ */
+function validateTypeImplementsInterface(
+  typeName: string,
+  typeFields: ObjectFieldInfo[],
+  interfaceSymbol: GraphQLOutputSymbol,
+): void {
+  const interfaceName = String(interfaceSymbol.name);
```

Comment:
If you follow my suggestion above for implementing `FieldInfoSet`, then this method becomes exceedingly simple:
```typescript
// typeFields, interfaceFields must be a FieldInfoSet
const interfaceFields = extractFieldInfo(interfaceSymbol);
return typeFields.isSupersetOf(interfaceFields);
```

If you want more specific error messages, you can break it up further. But you avoid a lot of verbosity and also a lot of iteration.